### DISCUSSION TITLE
How to know that action has new data?

### DISCUSSION THREAD

**Discussion by nemonemi** [remix-run/remix#7675] [Q&A] [answered]
I have a search form that after submitting returns a list. I handle the form submission in the Action, and it returns data correctly, however, when I change the search criteria, due to my UI being a 2-step form, I need to know when to go to the second step, otherwise, if I do it without regarding the data fetching, then a user might see the old data, before getting the updated one.

**Comment by kiliman** [comment]
Are you using `Form` or fetchers?

If you're using `Form`, I would recommend against returning *good* data from actions. You should only return *validation errors*. Instead, redirect to the correct route and let the loader return the data.

For fetchers, you can tell if data is *new* if `fetcher.state` goes from `idle` -> `loading|submitting` -> `idle`.

**Reply by nemonemi** [reply]
Hi @kiliman, thanks!
I'm using Form.

On that route, I already have one set of data being returned from the loader on that route -> **Projects**.
Then, when one interacts with a project, let's say to enhance its data, he searches for some records, let's call them **Items**.
When the search is performed (via action), and I "redirect" to a loader, how do I now return Items, and not Projects?

Also, what do you mean by "redirecting"? All of these queries and mutations happen on the same route.

**Reply by kiliman** [reply, marked as answer]
Do you have a simple example showing the route structure? It doesn't have to be real data.

For "search" forms, I recommend you use `method="get"` (default) and let your loader return the current items. The filter will be included in the search params.

```ts
// routes/projects/$projectId.tsx
export async function loader({ request, params }: DataFunctionArgs) {
  const url = new URL(request.url)
  const query = url.searchParams.get('query')
  
  const project = await getProject(params.projectId)
  let items = []
  if (query) {
    items = await getProjectItems(project, query)
  }
  return json({ project, items })
}

export default function Component() {
  const { project, items } = useLoaderData<typeof loader>()
  return (
    <div>
      <Form>
       <input type="text" name="query" />
      </Form>
      <pre>{JSON.stringify(product, null, 2)}</pre>
      <pre>{JSON.stringify(items, null, 2)}</pre>
    </div>
  )
}
```

**Reply by nemonemi** [reply]
That's a great tip on how to handle my two queries on the same route! Thanks!

However, it still doesn't resolve the issue I'm having, when I submit new search values, the component prints out first the old data, and then the new one.

Is there a way to know, without implementing the diff mechanism manually, that the data has changed?

Currently, this makes showing the UI different between "There are results" and "No! There aren't results" janky because if I search for something that exists, and then go back and search for something that doesn't, because it's a 2 step form (first form search, second form show) in the second form, when I show it, because it's unconnected to the server response, I show the response which already exists, and then it gets updated.

Here is my route structure
```ts
const router = createBrowserRouter(
  [
    {
      // Default URL - /projects
      path: '',
      loader: () => redirect(qualificationsUrl),
      element: null,
      errorElement: <ErrorScreen />,
      hasErrorBoundary: true,
    },
    {
      path: '/',
      loader: currentUserLoaderSetup(BACKEND_API_URL),
      element: <App />,
      id: 'currentUser',
      ErrorBoundary: AppErrorBoundary,
      hasErrorBoundary: true,
      children: [
        {
          path,
          loader: projectLoaderSetup(env.BACKEND_API_URL),
          action: projectActionSetup(env.BACKEND_API_URL),
          element: (<Project />),
          ErrorBoundary: () => <div>Error has occurred</div>,
          hasErrorBoundary: true,
        }
      ],
    },
    {
      path: '/error-screen',
      element: <ErrorScreen />,
      hasErrorBoundary: true,
    },
  ],
  { basename: BASE_PATH }
);
```

**Reply by nemonemi** [reply]
So, in regards to the behavior I was describing, I've resolved it by clearing the search query from the URL each time a user navigates to the first step of the form, with the search field.
 
Thanks for your prompt assistance @kiliman!
